### PR TITLE
New mocks

### PR DESCRIPTION
**Pull request by tonysimpson**

### CODE REVIEW

File: acceptable/mocks.py [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+    def url(self):
+        return self._url
+
+    @property
+    def methods(self):
+        return list(self._methods)
+
+    @property
+    def call_recorder(self):
+        return self._call_recorder
```
Comment (bloodearnest) [inline acceptable/mocks.py]: I don't think that properties are worth it here - why not just use regular members?

File: acceptable/mocks.py [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+
+    def _start(self):
+        _responses_manager.attach_callback(
+            self._methods, self._mock._url, self._mock._callback
+        )
+
+    def _stop(self):
+        _responses_manager.detach_callback(
+            self._methods, self._mock._url, self._mock._callback
+        )
```
Comment (bloodearnest) [inline acceptable/mocks.py]: Again, why not just .start() and .stop()? What does the underscore gain us - it is a public api really?

File: acceptable/mocks.py [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+    def name(self):
+        return self._name
+
+    @property
+    def url(self):
+        return self._url
+
+    @property
+    def methods(self):
+        return list(self._methods)
```
Comment (bloodearnest) [inline acceptable/mocks.py]: As above, regarding properties

File: acceptable/tests/test_mocks.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        with service() as mock:
+            with ExpectedException(AssertionError):
+                requests.get("http://example.com/test-endpoint", json="string")
+        assert_that(mock.endpoints.test_endpoint.get_calls(), HasLength(1))
+
+    def test_responses_manager_resets_responses_mock(self):
+        service = self.make_test_service()
+        with service() as service_mock:
+            with responses_mock_context() as responses_mock:
+                responses_mock.add('GET', 'http://exmaple.com/responses-test', b'test')
```
Comment (bloodearnest) [inline acceptable/tests/test_mocks.py]: Is "exmaple" a deliberate typo for the test? If so, maybe a comment or more obvious typo?

File: acceptable/mocks.py [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+            self._name,
+            self._url,
+            self._request_schema,
+            self._response_schema,
+            response_callback,
+        )
+
+
+class ServiceMock(object):
+    """User endpoints to endpoint mocks"""
```
Comment (bloodearnest) [inline acceptable/mocks.py]: I don't really understand this docstring?

File: acceptable/mocks.py [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+        try:
+            return self._attrs[name]
+        except KeyError:
+            raise NameError(name)
+
+    def __iter__(self):
+        return iter(self._attrs.items())
+
+
+class ResponsesManager(object):
```
Comment (bloodearnest) [inline acceptable/mocks.py]: Docstring explaining why we need a responses manager?

File: acceptable/mocks.py [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+from json import dumps as json_dumps
+from json import loads as json_loads
+from urllib.parse import urljoin
+import responses
+from acceptable._validation import validate
+import re
+from requests.utils import CaseInsensitiveDict
+
+
+class Attrs(object):
```
Comment (bloodearnest) [inline acceptable/mocks.py]: Docstring explaining rationale?

File: acceptable/mocks.py [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+
+    def get_calls(self):
+        return [c for m, c in self._calls]
+
+    def get_calls_for(self, mock):
+        return [c for m, c in self._calls if m == mock]
+
+    def get_calls_for_matching(self, mock, pattern):
+        if not hasattr(pattern, "search"):
+            pattern = re.compile(pattern)
```
Comment (bloodearnest) [inline acceptable/mocks.py]: Nice

File: acceptable/tests/test_mocks.py (hunk 4 above)
Comment (tonysimpson) [reply acceptable/tests/test_mocks.py]: nope just a typo

File: acceptable/mocks.py (hunk 5 above)
Comment (tonysimpson) [reply acceptable/mocks.py]: Neither do I, will fix

File: acceptable/mocks.py (hunk 6 above)
Comment (tonysimpson) [reply acceptable/mocks.py]: Good idea

File: acceptable/mocks.py (hunk 7 above)
Comment (tonysimpson) [reply acceptable/mocks.py]: good idea

File: acceptable/mocks.py (hunk 2 above)
Comment (tonysimpson) [reply acceptable/mocks.py]: these are not public

File: acceptable/mocks.py (hunk 1 above)
Comment (tonysimpson) [reply acceptable/mocks.py]: These are properties that should be immutable but are still useful to read. I think its useful to avoid misunderstanding.

**Pull request merged**